bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -398,7 +401,11 @@ synchronized long getLeastUnflushedLogId() {
     }
 
     synchronized long getCurrentLogId() {
-        return logChannel.getLogId();
+        return currentLogId;
+    }
+
+    BufferedLogChannel getCurrentLogChannel() {
```

Comment:
we can't just use `currentLogId` for testing the behavior of defer creation. defer creation means the `currentLogChannel` is null.

currentLogId and currentLogChannel are single entry log manager stuffs. they should not be part of the interface, that's what I suggested in our discussion, hiding all the log id stuffs in the implementation. If you follow that suggestion, the currentLogId and currentLogChannel should be just package-protected methods in single entry log manager, and the test EntryLogTest should be changed to test single entry log manager rather than entry logger.

Reply:
my basic question is why do we need one more method - getCurrentLogChannel? why cannt we manage your testing needs with just one method here

```	 
synchronized long getCurrentLogId() {
    if (logChannel == null) {
      // return some invalid/unassigned id
    } else {
      return logChannel.getLogId();
    }
}
```

Reply:
ideally EntryLogger isn't supposed to expose both currentLogId and currentLogChannel, but we made peace with the fact for currentLogId, but I'm not convenient with one more non-ideal decision.

Reply:
Current EntryLogger assumes it is single entry log implementation. These methods are package-protected methods (not public method), they are exposed for testing purpose.

----- 

If you are thinking from EntryLogManager perspective, these methods should be hidden to single entry log manager implementation. These methods should be only package-protected at single log manager implementation. 

If you see inconvenient here, that means your EntryLogManager interface in #1281 is not defining the right interfaces. `currentLogId` and `currentLogChannel` should only matter and be testing on single log entry manager. you shouldn't never consider it for per ledger entry manager. If you think and define your EntryLogManager interface in this way, I don't think `currentLogId` and `currentLogChannel` is not a problem to you.

Reply:
@reddycharan 

> my basic question is why do we need one more method - getCurrentLogChannel? why cannt we manage your testing needs with just one method here

why `getCurrentLogChannel` is a problem to you? I don't really understand your concern here. are you concerning the method making your interface complicate or what? 

as I said if you think this make your interface complicated, that mean your interface doesn't provide the right abstracted methods. The methods I am adding here is only for single entry log implementation, that says `currentLogId` and `currentLogChannel` should *ONLY* belong to single entry log implementation (which current EntryLogger is exactly implemented). 

After #1281, these two methods should *ONLY* moved to single entry log implementation. You shouldn't consider implement these two methods with PerLedgerEntryLogManager, otherwise that means your interface defines wrong methods.

Reply:
yeah these aren't public methods, but since there are packaged methods there is a possibility that classes in this package might use it, just like getcurrentlogid is called from SortedLedgerStorage, which I'm afraid of. For these classes, the type of the entrylogmanager should be transparent and at that time entrylogmanager for entrylogperledger should also provide that api.

But still I don't understand why do you need one more variable - currentLogId and one more method - getcurrentlogchannel, when your test cases are just checking if it is null or not null. It is extra burden to keep these vaiables in sync currentlogchannel and currentLogId and maintain one more method - getcurrentlogchannel (which I don't think should be exposed in the first place).

If you still feel strong about it, then let's have just currentLogChannel variable and getCurrentLogChannel method. We can remove currentLogId variable and it's corresponding getter. (for my entrylogmanager manager I would return null for this method).

Reply:
> just like getcurrentlogid is called from SortedLedgerStorage, 
> for my entrylogmanager for entrylogperledger I would return null for this method

as I pointed out at different places, log id stuffs are implementation specific and they should be hidden to implementation of EntryLogManager. they don't belong to the interface of EntryLogManager. I proposed in the other comment to have a method at EntryLogManager `write(SkipLIstFlusher ...)`, to write a skip list to entry logger, in this way you can hide the single-entrylog-manager implementation into this method without interfering with `getCurrentLogId`. so SortedLedgerStorage will not call `getCurrentLogId` any more but the tests can still examine `getCurrentLogId` through single-log-manager-implementation at white-box testing.

But again this is a discussion should be in #1281 rather than here.
 
> currentLogId and one more method - getcurrentlogchannel, when your test cases are just checking if it is null or not null. It is extra burden to keep these vaiables in sync currentlogchannel and currentLogId and maintain one more method 

That's a valid point to ask rather than from EntryLogManager perspective. I will remove `getCurrentLogChannel`.

Reply:
> That's a valid point to ask rather than from EntryLogManager perspective. I will remove getCurrentLogChannel.

Peace! thats my concern to begin with, having two variables and two getters when just one should be sufficient.